snuba/migrations/migrations.py

```diff
@@ -0,0 +1,91 @@
+import logging
+
+from datetime import datetime
+from enum import Enum
+from importlib import import_module
+from typing import Optional, Sequence, Tuple
+
+from snuba.clusters.cluster import ClickhouseClientSettings, get_cluster, CLUSTERS
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations.operations import Operation
+
+
+logger = logging.getLogger("snuba.migrations")
+
+LOCAL_TABLE_NAME = "migrations_local"
+DIST_TABLE_NAME = "migrations_dist"
+
+# Currently only local migrations are supported
+TABLE_NAME = LOCAL_TABLE_NAME
+
+
+class App(Enum):
+    SYSTEM = "system"
+    SNUBA = "snuba"
+
+
+class Status(Enum):
+    NOT_STARTED = "not_started"
+    IN_PROGRESS = "in_progress"
+    COMPLETED = "completed"
+
+
+class Migration:
+    is_dangerous: bool
+    dependencies: Optional[Sequence[Tuple[str, str]]]
```

Comment:
There isn't really a use case for skipping certain migrations now. It's supposed to provide the future ability to skip certain groups of migrations (such as those associated with optional features) that we may not want to run in certain environments.

Reply:
Then, since this is a hypothetical feature that would not require the old data structures to directly support it, I would discourage adding this additional complexity at this stage. If a migration is allowed to define dependencies we need to support them properly and we would not here, so I would just remove them, and stick with strict sequential dependency. I would add explicit dependencies when we introduce a use case that would use them and we provide the full support for them.

Reply:
I think there could be use cases for this fairly soon. One example I can think of is when we add system migrations (for example when we add the consumer_offsets column). The second system migration shouldn't depend on snuba migrations, it should only depend on the first system migration. But the following snuba migration should depend on this system migration as well as the previous snuba migration and therefore would have 2 dependencies.

Reply:
In such case can't we just execute the system migration in the order where it is, which is after the snuba migrations that come before that or is there any drawback in doing so? If we introduce this relationship now, we will have to take that into account when writing migrations. People are likely going to start adding dependencies, and this will be inconsistent because anyway we do not honor them. What id someone adds two migrations in a PR and they add a dependency from the first to the second that contradict the sequential order?

In practice, as long as we have a sequential order and that's the only order we follow, adding a dependency relationship seems just adding complexity for no gain. Or am I missing something?

Reply:
What do you mean by "contradict the sequential order"? There isn't any other sequential order that should exist. Are you talking about someone writing a circular dependency, like migration a depends on b and b depends on a? Dependencies are supposed to be the only way in which order is defined so we would have to always honor them.

Reply:
Wait, so I may have misunderstood this. Doesn't the 0001 in the migration file name define the order in any way ?